grapple/types/pages.py

```diff
@@ -169,6 +169,66 @@ class Meta:
         interfaces = (PageInterface,)
 
 
+def get_preview_page(token):
+    """
+    Get a preview page from a token.
+    """
+    token_params_list = token.split(":").pop(0)
+    token_params_kvstr = token_params_list.split(";")
+
+    params = {}
+    for arg in token_params_kvstr:
+        key, value = arg.split("=")
+        params[key] = value
+
+    """
+    Is there a more convenient way to get at the get_page_from_preview_token
+    method? 
+    """
+
+    """ 
+    We get id when previewing a page that has been saved. We're going to use it
+    to get the page so we can lookup it's class in order to call 
+    get_page_from_preview_token
+    """
+    id = params.get("id")
+    if id:
+        page = WagtailPage.objects.get(pk=id).specific
+        if page:
+            cls = type(page)
+            """ 
+            is there a particular version of wagtail that requires checking that
+            get_page_from_preview_token exists? is there a version of wagtail 
+            at which we can drop this conditional safely?
+            """
+            if hasattr(cls, "get_page_from_preview_token"):
+                """we assume that get_page_from_preview_token validates the token"""
+                return cls.get_page_from_preview_token(token)
+
+    """ 
+    we get parent_id and page_type when previewing a page that has not been
```

Comment:
parent_id is added for convenience. if I recall correctly we needed to vary things a bit when generating the token. When you have an ID that gives enough to go on.

Reply:
we could eliminate the if id: code path here if the page_type was always sent. It might not be as 'fast', but it would reduce the amount of code we need to maintain.